File: FWI/Python/make_hourly.py
import datetime
import argparse
from math import exp, pi, sin

import pandas as pd

import util

C_TEMP = {"c_alpha": 0.0, "c_beta": 2.75, "c_gamma": -1.9}
C_RH = {"c_alpha": 0.25, "c_beta": 2.75, "c_gamma": -2.0}
C_WIND = {"c_alpha": 1.0, "c_beta": 1.5, "c_gamma": -1.3}


def make_prediction(
    fcsts,
    c_alpha,
    c_beta,
    c_gamma,
    v="TEMP",
    change_at="SUNSET",
    min_value=float("-inf"),
    max_value=float("inf"),
    intervals=1,
    verbose=False,
):
    if verbose:
        print(f"Predicting {v} changing at {change_at}")
    fcsts = fcsts.copy()
    var_min = f"{v}_MIN"
    var_max = f"{v}_MAX"
    fcsts["TIME_MIN"] = fcsts["SUNRISE"] + c_alpha
    fcsts["TIME_MAX"] = fcsts["SOLARNOON"] + c_beta
    fcsts["VAR_MIN_TOM"] = pd.concat(
        [fcsts[var_min][1:], pd.Series([fcsts[var_min].iloc[-1]])]
    ).values
    # FIX: simplify in R code too
    # FIX: R code seems to use NA instead of a value at the end?
    fcsts["TIME_MIN_TOM"] = pd.concat(
        [fcsts["TIME_MIN"][1:], pd.Series([fcsts["TIME_MIN"].iloc[-1]])]
    ).values
    if 1 != len(fcsts["ID"].unique()):
        raise RuntimeError("Expected a single unique ID")
    hours = pd.DataFrame({"HR": range(24)})
    cross = pd.merge(fcsts[["DATE"]], hours, how="cross")
    minutes = pd.DataFrame({"MINUTE": range(0, 60, 60 // intervals)})
    cross = pd.merge(cross, minutes, how="cross")
    cross["ID"] = fcsts["ID"].iloc[0]
    cross["TIME"] = cross.apply(lambda row: row["HR"] + row["MINUTE"] / 60.0, axis=1)
    for column in ["HR", "MINUTE", "TIME", "MON", "YR"]:
        if column in fcsts.columns:
            del fcsts[column]
    cross = pd.merge(cross, fcsts, on=["ID", "DATE"])
    cross["START_DATE"] = cross.apply(
        lambda row: (pd.to_datetime(row["DATE"]) - datetime.timedelta(days=1)).strftime(
            "%Y-%m-%d"
        )
        if row["TIME"] <= row["TIME_MIN"]
        else row["DATE"],
        axis=1,
    )
    cross["HOUR_CURVE"] = cross.apply(
        lambda row: (row["TIME"] + 24)
        if row["TIME"] <= row["TIME_MIN"]
        else row["TIME"],
        axis=1,
    )
    cross["IS_RISING"] = cross.apply(
        lambda row: False
        if row["TIME"] <= row["TIME_MIN"]
        else row["TIME"] < row[change_at],
        axis=1,
    )
    cross["TIME_G_MIN"] = cross.apply(
        lambda row: row["TIME_MIN"]
        if row["TIME"] <= row["TIME_MIN"]
        else row["TIME_MIN_TOM"],
        axis=1,
    )
    cross["VAR_G_MIN"] = cross.apply(
        lambda row: row[var_min]
        if row["TIME"] <= row["TIME_MIN"]
        else row["VAR_MIN_TOM"],
        axis=1,
    )
    rising = cross[cross["IS_RISING"] == True][:]
    falling = cross[cross["IS_RISING"] == False][:]
    # figure out values before change_at
    rising["F_OR_G"] = rising.apply(
        lambda row: (row["HOUR_CURVE"] - row["TIME_MIN"])
        / (row["TIME_MAX"] - row["TIME_MIN"]),
        axis=1,
    )
    rising[v] = rising.apply(
        lambda row: (
            row[var_min] + (row[var_max] - row[var_min]) * sin((pi / 2) * row["F_OR_G"])
        ),
        axis=1,
    )
    rising["VAR_CHANGE"] = rising.apply(
        lambda row: (
            row[var_min]
            + (row[var_max] - row[var_min])
            * sin(
                (pi / 2)
                * (
                    (row[change_at] - row["TIME_MIN"])
                    / (row["TIME_MAX"] - row["TIME_MIN"])
                )
            )
        ),
        axis=1,
    )
    # need to figure out var value at change_at
    tmp = rising[["START_DATE", "VAR_CHANGE"]].drop_duplicates()
    # now merge change_at values into falling
    falling = pd.merge(falling, tmp, on=["START_DATE"])
    falling["F_OR_G"] = falling.apply(
        lambda row: (row["HOUR_CURVE"] - row[change_at])
        / (24 - row[change_at] + row["TIME_G_MIN"]),
        axis=1,
    )
    falling[v] = falling.apply(
        lambda row: row["VAR_G_MIN"]
        + (row["VAR_CHANGE"] - row["VAR_G_MIN"]) * exp(c_gamma * row["F_OR_G"]),
        axis=1,
    )
    # combine and sort everything
    out = pd.concat([rising, falling])
    out = out.sort_values(["DATE", "HR"])
    out["TIMESTAMP"] = out.apply(
        lambda row: pd.to_datetime(f'{row["DATE"]} {row["HR"]}:{row["MINUTE"]}:00'),
        axis=1,
    )
    out[v] = out[v].apply(lambda x: min(max(x, min_value), max_value))
    return out


def do_prediction(fcsts, row_temp, row_wind, row_rh, intervals=1, verbose=False):
    if verbose:
        print("Doing prediction")
    v_temp = make_prediction(
        fcsts,
        row_temp["c_alpha"],
        row_temp["c_beta"],
        row_temp["c_gamma"],
        "TEMP",
        "SUNSET",
        intervals=intervals,
        verbose=verbose,
    )
    v_wind = make_prediction(
        fcsts,
        row_wind["c_alpha"],
        row_wind["c_beta"],
        row_wind["c_gamma"],
        "WS",
        "SUNSET",
        min_value=0,
        intervals=intervals,
        verbose=verbose,
    )
    # FIX: R version only looks at HOUR so wouldn't work with intervals != 1
    t = v_temp[["ID", "TIMESTAMP", "DATE", "HR", "MINUTE", "LAT", "LONG", "TIMEZONE", "TEMP"]]
    w = v_wind[["ID", "TIMESTAMP", "WS"]]
    out = pd.merge(t, w)
    rh = make_prediction(
        fcsts,
        row_rh["c_alpha"],
        row_rh["c_beta"],
        row_rh["c_gamma"],
        "RH_OPP",
        intervals=intervals,
        min_value=0,
        max_value=1,
        verbose=verbose,
    )
    rh["RH"] = rh["RH_OPP"].apply(lambda rh_opp: 100 * (1 - rh_opp))
    rh = rh[["ID", "TIMESTAMP", "RH"]]
    out = pd.merge(out, rh)
    output = out[["ID", "LAT", "LONG", "TIMEZONE", "TIMESTAMP", "TEMP", "WS", "RH"]][:]
    if verbose:
        print("Assigning times")
    output["HR"] = output["TIMESTAMP"].apply(lambda x: x.hour)
    output["MINUTE"] = output["TIMESTAMP"].apply(lambda x: x.minute)
    if verbose:
        print("Converting date")
    output["DATE"] = output["TIMESTAMP"].apply(lambda x: x.date().strftime("%Y-%m-%d"))
    if verbose:
        print("Allocating rain")
    rain = fcsts[["DATE", "PREC"]][:]
    rain["HR"] = 7
    rain["MINUTE"] = 0
    if verbose:
        print("Merging")
    cmp = pd.merge(output, rain, how="outer")
    cmp["PREC"] = cmp["PREC"].fillna(0)
    if verbose:
        print("Calculating times")
    cmp["YR"] = cmp["TIMESTAMP"].apply(lambda x: x.year)
    cmp["MON"] = cmp["TIMESTAMP"].apply(lambda x: x.month)
    cmp["DAY"] = cmp["TIMESTAMP"].apply(lambda x: x.day)
    cmp["TIME"] = cmp.apply(lambda row: row["HR"] + row["MINUTE"] / 60.0, axis=1)
    if verbose:
        print("Done prediction")
    return cmp


##
# Convert daily min/max values stream to hourly values stream.
# Uses Beck & Trevitt method with default A/B/G values.
#
# @param     w              daily min/max values weather stream [lat, long, timezone, yr, mon, day, temp_min, temp_max, rh_min, rh_max, ws_min, ws_max, prec]
# @param     skip_invalid   if station year data non-sequential, skip and raise warning
# @param     verbose        whether to output progress messages
# @return                   hourly values weather stream [lat, long, timezone, yr, mon, day, hr, temp, rh, wind, prec]
def minmax_to_hourly_single(w, skip_invalid=False, verbose=False):
    r = w.copy()
    r.columns = map(str.upper, r.columns)
    if 1 != len(r["LAT"].unique()):
        raise RuntimeError("Expected a single LAT value for input weather")
    if 1 != len(r["LONG"].unique()):
        raise RuntimeError("Expected a single LONG value for input weather")
    if len(w["TIMEZONE"].unique()) != 1:
        raise RuntimeError("Expected a single UTC offset (timezone) each station year")
    had_id = "ID" in r.columns
    if not had_id:
        r["ID"] = "STN"
    elif 1 != len(r["ID"].unique()):
        raise RuntimeError("Expected a single ID value for input weather")
    if 1 != len(r["YR"].unique()):
        raise RuntimeError("Expected a single YEAR value for input weather")
    r["HR"] = 12
    r["TIMESTAMP"] = r.apply(
        lambda row: datetime.datetime(
            int(row["YR"]), int(row["MON"]), int(row["DAY"]), int(row["HR"])
        ),
        axis=1,
    )
    if not (util.is_sequential_days(r)):
        if skip_invalid:
            raise RuntimeWarning(
                f'{r["ID"].iloc[0]} for {r["YR"].iloc[0]}' +
                    ' - Expected input to be sequential daily weather'
            )
        return None
    orig_dates = pd.DataFrame(
        {
            "date": map(
                lambda x: x.strftime("%Y-%m-%d"),
                r["TIMESTAMP"].apply(lambda x: x.date()).unique(),
            )
        }
    )
    # duplicate start and end days to assume their values for one day before and after dataset
    yest = r.iloc[0][:]
    tom = r.iloc[-1][:]
    yest["TIMESTAMP"] = yest["TIMESTAMP"] - datetime.timedelta(days=1)
    tom["TIMESTAMP"] = tom["TIMESTAMP"] + datetime.timedelta(days=1)
    r = pd.concat([pd.DataFrame(data=[yest]), r, pd.DataFrame(data=[tom])])
    r["DATE"] = r["TIMESTAMP"].apply(lambda x: x.date())
    r["YR"] = r["TIMESTAMP"].apply(lambda x: x.year)
    r["MON"] = r["TIMESTAMP"].apply(lambda x: x.month)
    r["DAY"] = r["TIMESTAMP"].apply(lambda x: x.day)
    r["HR"] = r["TIMESTAMP"].apply(lambda x: x.hour)
    dates = r["TIMESTAMP"].unique()
    latitude = r["LAT"].iloc[0]
    longitude = r["LONG"].iloc[0]
    r = util.get_sunlight(r, get_solrad = False)
    r.columns = map(str.upper, r.columns)  # get_sunlight outputs lowercase columns
    # FIX: is solar noon just midpoint between sunrise and sunset?
    r["SOLARNOON"] = r.apply(
        lambda row: (row["SUNSET"] - row["SUNRISE"]) / 2 + row["SUNRISE"], axis=1
    )
    r["RH_OPP_MIN"] = 1 - r["RH_MAX"] / 100
    r["RH_OPP_MAX"] = 1 - r["RH_MIN"] / 100
    r["DATE"] = r["DATE"].apply(lambda x: x.strftime("%Y-%m-%d"))
    df = do_prediction(
        r, row_temp=C_TEMP, row_wind=C_WIND, row_rh=C_RH, verbose=verbose
    )
    df.columns = map(str.lower, df.columns)
    df = pd.merge(orig_dates, df, on=["date"])
    cols = ["lat", "long", "timezone", "yr", "mon", "day", "hr", "temp", "rh", "ws", "prec"]
    if had_id:
        df = df[["id"] + cols]
    else:
        df = df[cols]
    return df


# Convert daily min/max values stream to hourly values stream.
# Uses Beck & Trevitt method with default A/B/G values.
#
# @param    w               daily min/max values weather stream [lat, long, yr, mon, day temp_min, temp_max, rh_min, rh_max, ws_min, ws_max, prec]
# @param    timezone        UTC offset (default None for column provided in w)
# @param    skip_invalid    if station year data non-sequential, skip and raise warning
# @param    verbose         whether to output progress messages
# @param    round_out       decimals to truncate output to, None for none (default 4)
# @return                   hourly values weather stream [lat, long, timezone, yr, mon, day, hr, temp, rh, wind, prec]
def minmax_to_hourly(w, timezone = None, skip_invalid = False,
                     verbose = False, round_out = 4):
    r = w.copy()
    # check for required columns
    r.columns = map(str.upper, r.columns)
    req_cols = ["LAT", "LONG", "YR", "MON", "DAY", "TEMP_MIN", "TEMP_MAX",
                "RH_MIN", "RH_MAX", "WS_MIN", "WS_MAX", "PREC"]
    for col in req_cols:
        if not col in r.columns:
            raise RuntimeError("Missing required input column: " + col)
    # check for ID column
    had_id = "ID" in r.columns
    if not had_id:
        r["ID"] = "STN"
    # check timezone
    if timezone == None:
        if not "TIMEZONE" in r.columns:
            raise RuntimeError("Either provide a timezone column or specify argument in minmax_to_hourly")
    else:
        r["TIMEZONE"] = float(timezone)
    # loop over every station year
    result = None
    for stn in r["ID"].unique():
        by_stn = r[r["ID"] == stn]
        for yr in by_stn["YR"].unique():
            by_year = by_stn[by_stn["YR"] == yr]
            print(f"Running {stn} for {yr}")
            df = minmax_to_hourly_single(by_year, skip_invalid, verbose)
            result = pd.concat([result, df])

    # delete ID column if it wasn't provided
    if not had_id:
        del result["id"]

    # round decimal places of output columns
    if not (round_out == None or round_out == "None"):
        outcols = ["temp", "rh", "ws"]
        result[outcols] = result[outcols].map(round, ndigits = int(round_out))
    
    return result

if __name__ == "__main__":
    # run minmax_to_hourly by command line. run with option -h or --help to see usage
    parser = argparse.ArgumentParser(prog = "make_hourly")

    parser.add_argument("input", help = "Input csv data file")
    parser.add_argument("output", help = "Output csv file name and location")
    parser.add_argument("timezone", nargs = "?", default = None,
        help = "UTC offset (default None for column provided in input)")
    parser.add_argument("skip_invalid", nargs = "?", default = False,
        help = "If station year data non-sequential, skip and raise warning")
    parser.add_argument("-v", "--verbose", action = "store_true")
    parser.add_argument("-r", "--round_out", default = 4, nargs = "?",
        help = "Decimal places to truncate outputs to, None for no rounding (default 4)")

    args = parser.parse_args()
    df_in = pd.read_csv(args.input)
    df_out = minmax_to_hourly(df_in, args.timezone, args.skip_invalid,
                              args.verbose, args.round_out)
    df_out.to_csv(args.output, index = False)

File: FWI/R/make_hourly.r
# convert min/max weather format into hourly format
library(data.table)
library(lubridate)
source("util.r")

C_TEMP <- list(c_alpha = 0.0, c_beta = 2.75, c_gamma = -1.9)
C_RH <- list(c_alpha = 0.25, c_beta = 2.75, c_gamma = -2.0)
C_WIND <- list(c_alpha = 1.0, c_beta = 1.5, c_gamma = -1.3)

make_prediction <- function(fcsts, c_alpha, c_beta, c_gamma, v = "TEMP", change_at = "SUNSET",
  min_value = -Inf, max_value = Inf, intervals = 1, verbose = FALSE) {
  if (verbose) {
    print(paste0("Predicting ", v, " changing at ", change_at))
  }
  fcsts <- copy(fcsts)
  var_min <- paste0(v, "_MIN")
  var_max <- paste0(v, "_MAX")
  fcsts[, `:=`(
    TIME_MIN = SUNRISE + c_alpha,
    TIME_MAX = (SOLARNOON + c_beta),
    VAR_MIN_TOM = data.table::shift(get(var_min), -1),
    TIME_MIN_TOM = data.table::shift(SUNRISE, -1) + c_alpha
  )]
  # ~ fcsts[, `:=`(r2_time_min = data.table::shift(TIME_MIN, -1),
  # ~ r2_var_min = data.table::shift(get(var_min), -1))]
  stopifnot(1 == nrow(unique(fcsts$ID)))
  hours <- data.table(HR = 0:23)
  cross <- as.data.table(merge(as.data.frame(hours), as.data.frame(fcsts[, c("DATE")]),
    all = TRUE))
  minutes <- data.table(MINUTE = seq(0, 59, by = 60 / intervals))
  cross <- as.data.table(merge(as.data.frame(cross), as.data.frame(minutes), all = TRUE))
  cross[, ID := unique(fcsts$ID)]
  setorder(cross, cols = "DATE", "HR", "MINUTE")
  cross[, TIME := HR + MINUTE / 60.0]
  # cross <- merge(cross, fcsts[, -c('HR', 'MINUTE', 'TIME')], by=c('ID', 'DATE'))
  for (column in c("HR", "MINUTE", "TIME", "MON", "YR")) {
    if (column %in% colnames(fcsts)) {
      set(fcsts, , column, NULL)
    }
  }
  cross <- merge(cross, fcsts, by = c("ID", "DATE"))
  cross[, `:=`(
    START_DATE = ifelse(TIME <= TIME_MIN, as.character(as.Date(DATE) - 1), DATE),
    HOUR_CURVE = ifelse(TIME <= TIME_MIN, TIME + 24, TIME),
    IS_RISING = ifelse(TIME <= TIME_MIN, FALSE, TIME < get(change_at)),
    TIME_G_MIN = ifelse(TIME <= TIME_MIN, TIME_MIN, TIME_MIN_TOM),
    VAR_G_MIN = ifelse(TIME <= TIME_MIN, get(var_min), VAR_MIN_TOM)
  ), ]
  rising <- cross[IS_RISING == TRUE, ]
  falling <- cross[IS_RISING == FALSE, ]
  # figure out values before change_at
  rising[, F_OR_G := (HOUR_CURVE - TIME_MIN) / (TIME_MAX - TIME_MIN)]
  rising[, eval(v) := get(var_min) + (get(var_max) - get(var_min)) * sin((pi / 2) * F_OR_G)]
  rising[, VAR_CHANGE := get(var_min) + (get(var_max) - get(var_min)) *
    sin((pi / 2) * ((get(change_at) - TIME_MIN) / (TIME_MAX - TIME_MIN)))]
  # need to figure out var value at change_at
  tmp <- unique(rising[, c("START_DATE", "VAR_CHANGE")])
  # now merge change_at values into falling
  falling <- merge(falling, tmp, by = c("START_DATE"))
  falling[, F_OR_G := (HOUR_CURVE - get(change_at)) / (24 - get(change_at) + TIME_G_MIN)]
  falling[, eval(v) := VAR_G_MIN + (VAR_CHANGE - VAR_G_MIN) * exp(c_gamma * F_OR_G)]
  # combine and sort everything
  out <- rbind(rising, falling)
  setorder(out, cols = "DATE", "HR")
  out[, TIMESTAMP := as_datetime(paste0(as.character(DATE), " ", HR, ":", MINUTE, ":00"))]
  out[, eval(v) := pmin(pmax(get(v), min_value), max_value)]
  # HACK: somehow this isn't returning unless we do something to it first
  out <- out[!is.na(get(v))]
  return(copy(out))
}


do_prediction <- function(fcsts, row_temp, row_wind, row_RH, intervals = 1, verbose = FALSE) {
  if (verbose) {
    print("Doing prediction")
  }
  v_temp <- make_prediction(fcsts, row_temp$c_alpha, row_temp$c_beta, row_temp$c_gamma,
    "TEMP", "SUNSET", intervals = intervals, verbose = verbose)
  v_wind <- make_prediction(fcsts, row_wind$c_alpha, row_wind$c_beta, row_wind$c_gamma,
    "WS", "SUNSET", min_value = 0, intervals = intervals, verbose = verbose)
  t <- v_temp[, c("ID", "TIMESTAMP", "DATE", "HR", "LAT", "LONG", "TIMEZONE", "TEMP")]
  w <- v_wind[, c("ID", "TIMESTAMP", "DATE", "HR", "WS")]
  out <- merge(t, w)
  RH <- make_prediction(fcsts, row_RH$c_alpha, row_RH$c_beta, row_RH$c_gamma,
    "RH_OPP", intervals = intervals, min_value = 0, max_value = 1, verbose = verbose)
  RH[, `:=`(RH = 100 * (1 - RH_OPP))]
  RH <- RH[, c("ID", "TIMESTAMP", "RH")]
  out <- merge(RH, out, by = c("ID", "TIMESTAMP"))
  output <- out[, c("ID", "LAT", "LONG", "TIMEZONE", "TIMESTAMP", "TEMP", "WS", "RH")]
  # ~ output <- fwi(output)
  if (verbose) {
    print("Assigning times")
  }
  output[, HR := hour(TIMESTAMP)]
  output[, MINUTE := minute(TIMESTAMP)]
  if (verbose) {
    print("Converting date")
  }
  output[, DATE := as.character(as_date(TIMESTAMP))]
  if (verbose) {
    print("Allocating rain")
  }
  prec <- fcsts[, c("DATE", "PREC")]
  prec[, HR := 7]
  prec[, MINUTE := 0]
  if (verbose) {
    print("Merging")
  }
  cmp <- merge(output, prec, by = c("DATE", "HR", "MINUTE"), all = TRUE)[!is.na(TIMESTAMP)]
  cmp$PREC <- nafill(cmp$PREC, fill = 0)
  if (verbose) {
    print("Calculating times")
  }
  cmp[, YR := year(TIMESTAMP)]
  cmp[, MON := month(TIMESTAMP)]
  cmp[, DAY := day(TIMESTAMP)]
  cmp[, TIME := HR + MINUTE / 60.0]
  if (verbose) {
    print("Done prediction")
  }
  return(cmp)
}


#' Convert daily min/max values stream to hourly values stream.
#' Uses Beck & Trevitt method with default A/B/G values.
#'
#' @param     w         daily min/max values weather stream [lat, long, yr, mon, day, temp_min, temp_max, rh_min, rh_max, ws_min, ws_max, prec]
#' @return              hourly values weather stream [lat, long, timezone, yr, mon, day, hr, temp, rh, ws, prec]
minmax_to_hourly_single <- function(w, skip_invalid = FALSE, verbose = FALSE) {
  r <- copy(w)
  colnames(r) <- toupper(colnames(r))
  if (length(na.omit(unique(r$LAT))) != 1) {
    stop("Expected a single LAT value for input weather")
  }
  if (length(na.omit(unique(r$LONG))) != 1) {
    stop("Expected a single LONG value for input weather")
  }
  if (length(na.omit(unique(r$TIMEZONE))) != 1) {
    stop("Expected a single UTC offset (timezone) each station year")
  }
  timezone <- unique(r$TIMEZONE)
  hadId <- TRUE
  if (!("ID" %in% colnames(r))) {
    r$ID <- "STN"
    hadId <- FALSE
  } else if (length(na.omit(unique(r$ID))) != 1) {
    stop("Expected a single ID value for input weather")
  }
  if (length(na.omit(unique(r$YR))) != 1) {
    stop("Expected a single YR value for input weather")
  }
  r$HR <- 12
  # as_datetime() defaults to UTC, but we only use TIMESTAMP for it's combined yr, mon, day, hr
  r[, TIMESTAMP := make_datetime(YR, MON, DAY, HR)]
  if (!(nrow(r) == 1 || is_sequential(as.Date(r$TIMESTAMP)))) {
    if (skip_invalid) {
      warning(paste0(r$ID[[1]], " for ", r$YR[[1]],
        " - Expected input to be sequential daily weather"))
      return(NULL)
    }
    stop("Expected input to be sequential daily weather")
  }
  orig_dates <- data.table(date = as.character(unique(as_date(r$TIMESTAMP))))
  # duplicate start and end days to assume their values for one day before and after dataset
  yest <- r[1, ]
  tom <- r[nrow(r), ]
  yest[, TIMESTAMP := TIMESTAMP - days(1)]
  tom[, TIMESTAMP := TIMESTAMP + days(1)]
  r <- rbind(yest, r, tom)
  r[, DATE := as_date(TIMESTAMP)]
  r[, YR := year(TIMESTAMP)]
  r[, MON := month(TIMESTAMP)]
  r[, DAY := day(TIMESTAMP)]
  r[, HR := hour(TIMESTAMP)]
  r <- get_sunlight(r, get_solrad = FALSE)
  colnames(r) <- toupper(colnames(r))  # get_sunlight outputs lowercase columns
  # FIX: is solar noon just midpoint between sunrise and sunset?
  r[, SOLARNOON := (SUNSET - SUNRISE) / 2 + SUNRISE]
  r[, RH_OPP_MIN := 1 - RH_MAX / 100]
  r[, RH_OPP_MAX := 1 - RH_MIN / 100]
  r[, DATE := as.character(DATE)]
  df <- do_prediction(r, row_temp = C_TEMP, row_wind = C_WIND, row_RH = C_RH, verbose = verbose)
  colnames(df) <- tolower(colnames(df))
  df <- merge(orig_dates, df, by = "date")
  cols <- c("lat", "long", "timezone", "yr", "mon", "day", "hr", "temp", "rh", "ws", "prec")
  if (hadId) {
    df <- df[, c("id", ..cols)]
  } else {
    df <- df[, ..cols]
  }
  return(df)
}

#' Convert daily min/max values stream to hourly values stream.
#' Uses Beck & Trevitt method with default A/B/G values.
#'
#' @param   w           daily min/max values weather stream [lat, long, yr, mon, day, temp_min, temp_max, rh_min, rh_max, ws_min, ws_max, prec]
#' @param   timezone    UTC offset (default NA for column provided in w)
#' @param   skipInvalid if station year data non-sequential, skip and raise warning
#' @param   verbose     whether to output progress messages
#' @param   round_out   decimals to truncate output to, NA for none (default 4)
#' @return              hourly values weather stream [lat, long, timezone, yr, mon, day, hr, temp, rh, ws, prec]
#' @export minmax_to_hourly
minmax_to_hourly <- function(w, timezone = NA, skip_invalid = FALSE,
  verbose = FALSE, round_out = 4) {
  # check df_wx class for data.frame or data.table
  wasDT <- is.data.table(w)
  if (wasDT) {
    r <- copy(w)
  } else if (is.data.frame(w)) {
    r <- copy(w)
    setDT(r)
  } else {
    stop("Input weather stream w needs to be a data.frame or data.table!")
  }

  # check for required columns
  colnames(r) <- toupper(colnames(r))
  req_cols <- c("LAT", "LONG", "YR", "MON", "DAY", "TEMP_MIN", "TEMP_MAX",
                "RH_MIN", "RH_MAX", "WS_MIN", "WS_MAX", "PREC")
  for (col in req_cols) {
    if (!col %in% names(r)) {
      stop(paste("Missing required input column:", col))
    }
  }
  # check for ID column
  hadId <- TRUE
  if (!("ID" %in% colnames(r))) {
    r$ID <- "STN"
    hadId <- FALSE
  }
  # check timezone
  if (is.na(timezone) || timezone == "NA") {
    if (!"TIMEZONE" %in% names(r)) {
      stop("Either provide a timezone column or specify argument in minmax_to_hourly")
    }
  } else {
    r[, TIMEZONE := as.numeric(..timezone)]
  }
  # loop over every station year
  result <- NULL
  for (stn in unique(r$ID)) {
    by_stn <- r[ID == stn, ]
    for (yr in unique(by_stn$YR)) {
      by_year <- by_stn[YR == yr, ]
      print(paste0("Running ", stn, " for ", yr))
      df <- minmax_to_hourly_single(by_year, skip_invalid, verbose)
      result <- rbind(result, df)
    }
  }

  # remove ID column if it wasn't provided
  if (!hadId) {
    result[, id := NULL]
  }

  # format decimal places of output columns
  if (!(is.na(round_out) || round_out == "NA")) {
    outcols <- c("temp", "rh", "ws")
    set(result, j = outcols, value = round(result[, ..outcols], as.integer(round_out)))
  }

  if (!wasDT) {
    setDF(result)
  }
  return(result)
}

# run minmax_to_hourly by command line via Rscript, requires 2 args: input csv and output csv
# optional args: timezone, skip_invalid, verbose, round_out
if ("--args" %in% commandArgs() && sys.nframe() == 0) {
  args <- commandArgs(trailingOnly = TRUE)
  if (length(args) < 2) {
    stop("at least 2 arguments required: input csv and output csv")
  }
  input <- args[1]
  output <- args[2]
  # load optional arguments if provided, or set to default
  if (length(args) >= 3) timezone <- as.numeric(args[3])
  else timezone <- NA
  if (length(args) >= 4) skip_invalid <- as.numeric(args[3])
  else skip_invalid <- FALSE
  if (length(args) >= 5) verbose <- as.logical(args[4])
  else verbose <- FALSE
  if (length(args) >= 6) round_out <- args[5]
  else round_out <- 4
  if (length(args) >= 7) warning("Too many input arguments provided, some unused")

  df_in <- read.csv(input)
  df_out <- minmax_to_hourly(df_in, timezone, skip_invalid, verbose, round_out)
  write.csv(df_out, output, row.names = FALSE)
}
